smac/intensification/intensification.py

```diff
@@ -137,30 +192,303 @@ def __init__(self, tae_runner: ExecuteTARun,
+
+
+        Parameters
+        ----------
+        challengers : typing.List[Configuration]
+            promising configurations
+        incumbent: Configuration
+            incumbent configuration
+        chooser : smac.optimizer.epm_configuration_chooser.EPMChooser
+            optimizer that generates next configurations to use for racing
+        run_history : typing.Optional[smac.runhistory.runhistory.RunHistory]
+            stores all runs we ran so far
+        repeat_configs : bool
+            if False, an evaluated configuration will not be generated again
+
+        Returns
+        -------
+        behest: IntensifierBehest
+            What should the smbo object do with the runinfo.
+        run_info: RunInfo
+            An object that encapsulates necessary information for a config run
+        """
+
+        # In case a crash happens, and FirstRunCrashedException prevents a
+        # failure, revert back to running the incumbent
+        # Challenger case is by construction ok, as there is no special
+        # stage for its processing
+        if self.stage == IntensifierStage.PROCESS_FIRST_CONFIG_RUN:
+            self.stage = IntensifierStage.RUN_FIRST_CONFIG
+        elif self.stage == IntensifierStage.PROCESS_INCUMBENT_RUN:
+            self.stage = IntensifierStage.RUN_INCUMBENT
+
+        try:
+
+            # Line 2:
+            # Understand who is the challenger.
+            # Not necessarily a challenger is always a new configuration,
+            # as this method might be called multiple times to
+            # gain more confidence on incumbent, or even run the same
+            # challenger on multiple N (line 17)
+            challenger = self.query_challenger(
+                challengers=challengers,
+                chooser=chooser,
+                run_history=run_history,
+                repeat_configs=repeat_configs,
+            )
+
+            # if first ever run, then assume current challenger to be the incumbent
+            # In this case, we need a branch new challenger
+            if self.stage == IntensifierStage.RUN_FIRST_CONFIG:
+                if incumbent is None:
+                    self.logger.info(
+                        "First run, no incumbent provided;"
+                        " challenger is assumed to be the incumbent"
+                    )
+                    incumbent = challenger
+                else:
+                    inc_runs = run_history.get_runs_for_config(
+                        incumbent,
+                        only_max_observed_budget=True
+                    )
+                    if len(inc_runs) > 0:
+                        self.logger.debug(
+                            "Skipping RUN_FIRST_CONFIG stage since "
+                            "incumbent is already run before"
+                        )
+                        self.stage = IntensifierStage.RUN_INCUMBENT
+
+            # Before we proceed with the intensification loop,
```

Comment:
Is this here only for optimization? If yes, I would like to leave it out of this PR to keep the intensification code as simple as possible for now.

Reply:
Wait, I think the biggest part of the complexity overhead is to remove the None scenario.

What I mean by this, is that the previous code outputs None as a criteria to skip an iteration/move to a new iteration and so on. 

I thought we wanted to remove this behavior and always output a valid configuration for a runner to launch. Hence we look ahead of time if there are not gonna be challengers, instances and so on.

If that is not a requirement, then the code can be greatly simplified by removing not only this if but many other look-ahead places in the code. Please let me know.

Reply:
> If that is not a requirement, then the code can be greatly simplified by removing not only this if but many other look-ahead places in the code. Please let me know.

In my current understanding we don't require intensify to return a valid runinfo as we return the `IntensificationBehest` to allow skipping a configuration/having the function called again. I guess this will make it a bit more inefficient, but it will make the code much more readable and simpler for now. Does this make sense?

Reply:
a lot, thanks, this will allow further simplification. Let me remove some code here and there and do another push.